File: examples/atomic-inc/go/atomic_increment.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"

	"github.com/joho/godotenv"
)

var (
	hpkvBaseURL string
	hpkvAPIKey  string
)

func init() {
	// Load environment variables from .env file
	if err := godotenv.Load(); err != nil {
		fmt.Printf("Warning: Error loading .env file: %v\n", err)
	}

	hpkvBaseURL = os.Getenv("HPKV_BASE_URL")
	hpkvAPIKey = os.Getenv("HPKV_API_KEY")
}

func createKey(key string, initialValue int) (bool, error) {
	payload := map[string]string{
		"key":   key,
		"value": fmt.Sprintf("%d", initialValue),
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return false, fmt.Errorf("error marshaling payload: %v", err)
	}

	fmt.Printf("Creating key with payload: %s\n", string(jsonData))

	req, err := http.NewRequest("POST", fmt.Sprintf("%s/record", hpkvBaseURL), bytes.NewBuffer(jsonData))
	if err != nil {
		return false, fmt.Errorf("error creating request: %v", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", hpkvAPIKey)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return false, fmt.Errorf("error making request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return false, fmt.Errorf("error reading response body: %v", err)
	}

	fmt.Printf("Create response status: %d\n", resp.StatusCode)
	fmt.Printf("Create response body: %s\n", string(body))

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return false, fmt.Errorf("create failed with status %d - %s", resp.StatusCode, string(body))
	}

	return true, nil
}

func atomicIncrement(key string, increment int) (map[string]interface{}, error) {
	if hpkvBaseURL == "" || hpkvAPIKey == "" {
		return nil, fmt.Errorf("HPKV_BASE_URL and HPKV_API_KEY must be set in environment variables")
	}

	payload := map[string]interface{}{
		"key":       key,
		"increment": increment,
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("error marshaling payload: %v", err)
	}

	fmt.Printf("Attempting atomic increment with payload: %s\n", string(jsonData))

	req, err := http.NewRequest("POST", fmt.Sprintf("%s/record/atomic", hpkvBaseURL), bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("error creating request: %v", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", hpkvAPIKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading response body: %v", err)
	}

	fmt.Printf("Atomic increment response status: %d\n", resp.StatusCode)
	fmt.Printf("Atomic increment response body: %s\n", string(body))

	// If the key doesn't exist (404), create it first and try again
	if resp.StatusCode == http.StatusNotFound {
		fmt.Printf("Key '%s' doesn't exist. Creating it with initial value 0...\n", key)
		if success, err := createKey(key, 0); !success {
			return nil, fmt.Errorf("failed to create key with initial value: %v", err)
		}

		// Retry the increment operation
		fmt.Println("Retrying atomic increment after key creation...")
		req, err = http.NewRequest("POST", fmt.Sprintf("%s/record/atomic", hpkvBaseURL), bytes.NewBuffer(jsonData))
		if err != nil {
			return nil, fmt.Errorf("error creating retry request: %v", err)
		}

		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("x-api-key", hpkvAPIKey)

		resp, err = client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("error making retry request: %v", err)
		}
		defer resp.Body.Close()

		body, err = io.ReadAll(resp.Body)
		if err != nil {
			return nil, fmt.Errorf("error reading retry response body: %v", err)
		}

		fmt.Printf("Retry response status: %d\n", resp.StatusCode)
		fmt.Printf("Retry response body: %s\n", string(body))
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("error unmarshaling response: %v", err)
	}

	if success, ok := result["success"].(bool); !ok || !success {
		message := "Unknown error"
		if msg, ok := result["message"].(string); ok {
			message = msg
		}
		return nil, fmt.Errorf("HPKV API error: %s", message)
	}

	return result, nil
}

func main() {
	key := "counter:example"

	// Clean up any existing key
	fmt.Println("Cleaning up any existing key...")
	req, err := http.NewRequest("DELETE", fmt.Sprintf("%s/record/%s", hpkvBaseURL, key), nil)
	if err != nil {
		fmt.Printf("Error creating delete request: %v\n", err)
		os.Exit(1)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", hpkvAPIKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Error making delete request: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("Error reading delete response body: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Delete response status: %d\n", resp.StatusCode)
	fmt.Printf("Delete response body: %s\n", string(body))

	if _, err := createKey(key, 0); err != nil {
		fmt.Printf("Error creating key: %v\n", err)
		os.Exit(1)
	}

	// Get the initial value
	getReq, err := http.NewRequest("GET", fmt.Sprintf("%s/record/%s", hpkvBaseURL, key), nil)
	if err != nil {
		fmt.Printf("Error creating get request: %v\n", err)
		os.Exit(1)
	}

	getReq.Header.Set("Content-Type", "application/json")
	getReq.Header.Set("x-api-key", hpkvAPIKey)

	getResp, err := client.Do(getReq)
	if err != nil {
		fmt.Printf("Error making get request: %v\n", err)
		os.Exit(1)
	}
	defer getResp.Body.Close()

	fmt.Printf("Get response status: %d\n", getResp.StatusCode)

	// Increment by 1
	fmt.Println("\nIncrementing counter by 1...")
	result, err := atomicIncrement(key, 1)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Result: %+v\n", result)

	// Increment by 5
	fmt.Println("\nIncrementing counter by 5...")
	result, err = atomicIncrement(key, 5)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Result: %+v\n", result)

	// Decrement by 2
	fmt.Println("\nDecrementing counter by 2...")
	result, err = atomicIncrement(key, -2)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Result: %+v\n", result)
}

File: examples/atomic-inc/csharp/AtomicIncrement.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Extensions.Configuration;

class Program
{
    private static readonly string? HpkvBaseUrl;
    private static readonly string? HpkvApiKey;

    static Program()
    {
        var configuration = new ConfigurationBuilder()
            .SetBasePath(Directory.GetCurrentDirectory())
            .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
            .AddEnvironmentVariables()
            .Build();

        HpkvBaseUrl = configuration["HPKV_BASE_URL"];
        HpkvApiKey = configuration["HPKV_API_KEY"];
    }

    private static async Task<bool> CreateKey(string key, int initialValue = 0)
    {
        try
        {
            var payload = new
            {
                key = key,
                value = initialValue.ToString()
            };

            Console.WriteLine($"Creating key with payload: {JsonSerializer.Serialize(payload)}");

            using var client = new HttpClient();
            client.DefaultRequestHeaders.Add("x-api-key", HpkvApiKey);

            var content = new StringContent(
                JsonSerializer.Serialize(payload),
                Encoding.UTF8,
                "application/json"
            );

            var response = await client.PostAsync($"{HpkvBaseUrl}/record", content);
            var responseContent = await response.Content.ReadAsStringAsync();

            Console.WriteLine($"Create response status: {(int)response.StatusCode}");
            Console.WriteLine($"Create response body: {responseContent}");

            if (!response.IsSuccessStatusCode)
            {
                Console.WriteLine($"Create failed with status {(int)response.StatusCode} - {responseContent}");
                return false;
            }

            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error creating record: {ex.Message}");
            return false;
        }
    }

    private static async Task<JsonDocument> AtomicIncrement(string key, int increment)
    {
        if (string.IsNullOrEmpty(HpkvBaseUrl) || string.IsNullOrEmpty(HpkvApiKey))
        {
            throw new ArgumentException("HPKV_BASE_URL and HPKV_API_KEY must be set in environment variables");
        }

        try
        {
            var payload = new
            {
                key = key,
                increment = increment
            };

            Console.WriteLine($"Attempting atomic increment with payload: {JsonSerializer.Serialize(payload)}");

            using var client = new HttpClient();
            client.DefaultRequestHeaders.Add("x-api-key", HpkvApiKey);

            var content = new StringContent(
                JsonSerializer.Serialize(payload),
                Encoding.UTF8,
                "application/json"
            );

            var response = await client.PostAsync($"{HpkvBaseUrl}/record/atomic", content);
            var responseContent = await response.Content.ReadAsStringAsync();

            Console.WriteLine($"Atomic increment response status: {(int)response.StatusCode}");
            Console.WriteLine($"Atomic increment response body: {responseContent}");

            if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            {
                Console.WriteLine($"Key '{key}' doesn't exist. Creating it with initial value 0...");
                if (!await CreateKey(key, 0))
                {
                    throw new Exception("Failed to create key with initial value");
                }

                Console.WriteLine("Retrying atomic increment after key creation...");
                response = await client.PostAsync($"{HpkvBaseUrl}/record/atomic", content);
                responseContent = await response.Content.ReadAsStringAsync();

                Console.WriteLine($"Retry response status: {(int)response.StatusCode}");
                Console.WriteLine($"Retry response body: {responseContent}");
            }

            response.EnsureSuccessStatusCode();
            var result = JsonDocument.Parse(responseContent);

            if (!result.RootElement.GetProperty("success").GetBoolean())
            {
                var errorMsg = result.RootElement.GetProperty("message").GetString() ?? "Unknown error";
                throw new Exception($"HPKV API error: {errorMsg}");
            }

            return result;
        }
        catch (HttpRequestException ex)
        {
            throw new Exception($"Failed to connect to HPKV API: {ex.Message}");
        }
        catch (Exception ex)
        {
            throw new Exception($"Invalid response from HPKV API: {ex.Message}");
        }
    }

    static async Task Main(string[] args)
    {
        var key = "counter:example";

        try
        {
            // Clean up any existing key
            Console.WriteLine("Cleaning up any existing key...");
            using var client = new HttpClient();
            client.DefaultRequestHeaders.Add("x-api-key", HpkvApiKey);

            var deleteResponse = await client.DeleteAsync($"{HpkvBaseUrl}/record/{key}");
            var deleteContent = await deleteResponse.Content.ReadAsStringAsync();
            Console.WriteLine($"Delete response status: {(int)deleteResponse.StatusCode}");
            Console.WriteLine($"Delete response body: {deleteContent}");

            await CreateKey(key, 0);

            var getResponse = await client.GetAsync($"{HpkvBaseUrl}/record/{key}");
            Console.WriteLine($"Get response status: {(int)getResponse.StatusCode}");

            // Increment by 1
            Console.WriteLine("\nIncrementing counter by 1...");
            var result = await AtomicIncrement(key, 1);
            Console.WriteLine($"Result: {result.RootElement.GetRawText()}");

            // Increment by 5
            Console.WriteLine("\nIncrementing counter by 5...");
            result = await AtomicIncrement(key, 5);
            Console.WriteLine($"Result: {result.RootElement.GetRawText()}");

            // Decrement by 2
            Console.WriteLine("\nDecrementing counter by 2...");
            result = await AtomicIncrement(key, -2);
            Console.WriteLine($"Result: {result.RootElement.GetRawText()}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }
} 